### DISCUSSION TITLE
Issue importing micronaut 3 library into micronaut 4 application

### DISCUSSION THREAD

**Discussion by pedrolamarao** [micronaut-projects/micronaut-core#9758] [Q&A]
Hello all! We have a micronaut 3 library providing bean factories to ease integrating our products into micronaut applications. Micronaut 3 applications can inject beans provided by our bean factories just fine. Micronaut 4 application cannot, throwing:

`Caused by: io.micronaut.context.exceptions.NoSuchBeanException: No bean of type [library.Foo] exists. Make sure the bean is not disabled by bean requirements (enable trace logging for 'io.micronaut.context.condition' to check) and if the bean is enabled then ensure the class is declared a bean and annotation processing is enabled (for Java and Kotlin the 'micronaut-inject-java' dependency should be configured as an annotation processor).`

We have studied the migration guide and compatibility break documentation to no avail.

The following test application demonstrates our troubles. While :application3 endpoints complete sucessfully, :application4 endpoints all fail as above. Are we missing configuration? Please advise!

https://github.com/pedrolamarao/micronaut-library-issue

**Comment by dstepanov** [comment]
The library needs to be recompiled with v4

**Reply by pedrolamarao** [reply]
Thank you. I have not seen mention of this in any upgrade guides.

**Reply by hrothwell** [reply]
@dstepanov This should really be something mentioned in the upgrade guides as @pedrolamarao mentioned, this can be a serious blocker for upgrading for teams that may depend on external micronaut libraries.

For example, a team within our enterprise creates and maintains several libraries utilizing micronaut that help speed up the workflow of other teams using micronaut. Our team is one such team and spent several days trying to figure out what was going wrong with our upgrade until we got lucky and stumbled across this git issue. Now we at least know we are dependent on the other team upgrading first, but this could have saved a lot of time to have a quick message in the breaking changes / upgrade notes for Micronaut 4.

**Reply by hrothwell** [reply]
https://github.com/micronaut-projects/micronaut-core/pull/10248
